neurolib/models/multimodel/builder/adex.py

```diff
@@ -0,0 +1,944 @@
+        """
+        # load linear-nonlinear cascade quantities from file
+        logging.info(f"Reading precomputed quantities from {filename}")
+        loaded_h5 = File(filename, "r")
+        self.mu_range = np.array(loaded_h5["mu_vals"])
+        self.d_mu = self.mu_range[1] - self.mu_range[0]
+        self.sigma_range = np.array(loaded_h5["sigma_vals"])
+        self.d_sigma = self.sigma_range[1] - self.sigma_range[0]
+        self.firing_rate_cascade = np.array(loaded_h5["r_ss"])
+        self.voltage_cascade = np.array(loaded_h5["V_mean_ss"])
+        self.tau_cascade = np.array(loaded_h5["tau_mu_exp"])
+        logging.info("Loading done, all quantities loaded.")
+        # close the file
+        loaded_h5.close()
+        self.lin_nonlin_fname = filename
+
+    def update_params(self, params_dict):
+        """
+        Update parameters as in the base class but also rescale.
+        """
+        # if we are changing C_m or g_L, update tau_m as well
+        if any(k in params_dict for k in ("C_m", "g_L")):
+            C_m = params_dict["C_m"] if "C_m" in params_dict else self.params["C_m"]
+            g_L = params_dict["g_L"] if "g_L" in params_dict else self.params["g_L"]
+            params_dict["tau_m"] = C_m / g_L
+
+        # if we are chaning any of the J_exc_max, tau_syn_exc or c_global, rescale c_global
+        if any(k in params_dict for k in ("c_global", "J_exc_max", "tau_syn_exc")):
+            # get original c_global
+            c_global = (
+                params_dict["c_global"]
+                if "c_global" in params_dict
+                else self.params["c_global"] * (self.params["J_exc_max"] / self.params["tau_syn_exc"])
+            )
+            tau_syn_exc = params_dict["tau_syn_exc"] if "tau_syn_exc" in params_dict else self.params["tau_syn_exc"]
+            J_exc_max = params_dict["J_exc_max"] if "J_exc_max" in params_dict else self.params["J_exc_max"]
+            params_dict["c_global"] = c_global * tau_syn_exc / J_exc_max
+
+        # update all parameters finally
+        super().update_params(params_dict)
+
+    def describe(self):
+        return {
+            **super().describe(),
+            "lin_nonlin_cascade_filename": self.lin_nonlin_fname,
+        }
+
+    def _callbacks(self):
+        """
+        Construct list of python callbacks for AdEx model.
+        """
+        callbacks_list = [
+            (self.callback_functions["firing_rate_lookup"], self.firing_rate_lookup, 2),
+            (self.callback_functions["voltage_lookup"], self.voltage_lookup, 2),
+            (self.callback_functions["tau_lookup"], self.tau_lookup, 2),
+        ]
+        self._validate_callbacks(callbacks_list)
+        return callbacks_list
+
+    def _numba_callbacks(self):
+        """
+        Define numba callbacks - has to be different than jitcdde callbacks
+        because of the internals.
+        """
+
+        def _table_numba_gen(sigma_range, mu_range, d_sigma, d_mu, cascade):
```

Comment:
this is a little trick where instead of something like
```python
partial(_table_lookup, sigma_range=..., mu_range=..., ...)
```
which cannot be wrapped in `numba.njit()` you do nested function and when you call it, you can still use `numba.njit()`... so this is really just `partial()` with the possibility of using `numba.njit()` lates

Reply:
ok 🤯